Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h

```diff
@@ -1022,9 +1035,21 @@ AirIRGeneratorBase<Derived, ExpressionType>::AirIRGeneratorBase(const ModuleInfo
                 // stack that such a stub would use.
                 Checked<uint32_t>(m_maxNumJSCallArguments) * sizeof(Register) + jsCallingConvention().headerSizeInBytes
             ));
-            const int32_t checkSize = m_makesCalls ? (wasmFrameSize + extraFrameSize).value() : wasmFrameSize.value();
+            int32_t checkSize = wasmFrameSize.value();
+            bool frameSizeNeedsOverflowCheck = checkSize >= static_cast<int32_t>(minimumParentCheckSize);
+            bool needsOverflowCheck = frameSizeNeedsOverflowCheck;
+
+            if (m_makesCalls) {
+                needsOverflowCheck = true;
+                checkSize = checkedSum<int32_t>(checkSize, extraFrameSize).value();
+            } else if (m_makesTailCalls) {
+                Checked<int32_t> tailCallCheckSize = std::max<Checked<int32_t>>(wasmTailCallFrameSize + extraFrameSize, 0);
+                checkSize = frameSizeNeedsOverflowCheck ? std::max<Checked<int32_t>>(tailCallCheckSize, wasmFrameSize).value() : tailCallCheckSize.value();
```

Comment:
> Shouldn't we add wasmFrameSize here?

I think it is now handled by `checkSize = frameSizeNeedsOverflowCheck ? max(tailCallCheckSize,wasmFrameSize) : tailCallCheckSize`
For reference:

We are discussing the specific case where we don't make any calls but we might tail call:

```
                   +----------------------------+
                   +                            +
                   +                            +
                   +                            +
                   +                            +
                   +                            +
                   +                            +
lower address      +----------FP----------------+  fp at function prologue, all size are calculated from here
              |    |                            |
              |    +                            +
              v    |                            |
                   +                            +
                   |--wasmTailCallFrameSize(+)--| where sp will be just before jumping to the callee
                   +                            +  ^
                   |                            |  |
                   +------wasmFrameSize---------+  |
                   |                            |  |
                   |                            |  | extraSize
                   |                            |  |
                   |--minimumParentCheckSize----|  |
                   +                            +  |              
                   |                            |  |             
                   +                            +  |            
                   |                            |  v           
                   +wasmTailCallFrameSize + extraSize
                   |                            |         
                   +                            +
                   |                            |       
                   +                            +
                   |                            |     
                   +                            +    
                   |                            |   
                   +                            +  
                   |                            | 
                   +                            |
                   |                            | 
                   +----------------------------|
```

Our caller did a check for `minimumParentCheckSize` and the total frame size is within this limit (`frameSizeNeedsOverflowCheck == false`).
However, we need to ensure the callee that we did the overflow check, so we add extra size to `wasmTailCallFrameSize`. In the diagram above, we see that we need an overflow check.
As mentioned, if we only tail call but the total frame size is bigger than `minimumParentCheckSize` (it would be below `minimumParentCheckSize` on diagram 1), then we do the overflow check with` checkSize = max(tailCallFrameSize + extraSize, frameSize)`.


Another case:

```



                   +----------------------------+
                   +                            +
                   +                            +
                   |--wasmTailCallFrameSize(-)--| where sp will be just before jumping to the callee
                   +                            + ^
                   +                            + |  
                   +                            + |
lower address      +----------FP----------------+ | 
              |    |                            | | extraSize
              |    +                            + |
              v    |                            | |
                   +                            + |
                   +                            + |  
                   |                            | v
                   +wasmTailCallFrameSize + extraSize
                   |                            |
                   +------wasmFrameSize---------+
                   |                            |
                   |                            |
                   |                            |
                   |--minimumParentCheckSize----|
                   +                            +
                   |                            |
                   +                            +
                   |                            |
                   |                            |         
                   +                            +
                   |                            |       
                   +                            +
                   |                            |     
                   +                            +    
                   |                            |   
                   +                            +  
                   |                            | 
                   +                            |
                   |                            | 
                   +----------------------------|
```

Here, the callee's stack args fit in the caller's stack arguments slots, so the stack pointer will be above current FP when tail-calling.
We don't need to do an overflow check.

> We can never reduce the frame size that we check against, a negative value here only means that none of our callees will ever have to do their own stack check.

The comment was misleading, I removed it.